File: TPII-2025-2/PadraoGoF/Estrutural/Adapter01.js
// LINK DO COD: https://abre.ai/fatec-tpii
// ********************************************************

// 1 - Interface do Cliente
class AplicacaoCliente{
    request() {}
}

class Cliente{
    constructor(target){
        this.target = target;
    }

    criaRequest(){
        console.log("Cliente - Fazendo uma requisição");
        this.target.request();
    }
}

// 2 - Serviço Existente
class Adaptee{
    especificaRequest(){
        console.log("requisição específica do Adaptee");
    }
}

// 3 - Adapter
class Adapter{
    constructor(adaptee){
        this.adaptee = adaptee;
    }

    request(){
        this.adaptee.especificaRequest();
    }
}


// 5 - UTILIZANDO O ADAPTAR -----------------------------------
const adaptee = new Adaptee();
const adapter = new Adapter(adaptee);
const cliente = new Cliente(adapter);

// Fazendo uma requisição - Requisição Especifica do Adaptee:
cliente.criaRequest();

File: TPII-2025-2/PadraoGoF/Estrutural/Adapter01.py
# interface do cliente
class AplicacaoCliente:
    def requisitar(self):
        pass

class Cliente:
    def __init__(self, target):
        self.target = target

    def criaRequest(self):
        print("Cliente: Fazendo uma requisição ")
        self.target.request()

# serviço existente
class Adaptee:
    def especificaRequest(self):
        print("Requisição especifica do Adaptee")

# Adapter
class Adapter(AplicacaoCliente):
    def __init__(self, adaptee):
        self.adaptee = adaptee
    
    def request(self):
        self.adaptee.especificaRequest()


# utilizando o adaptar
adaptee = Adaptee()
adapter = Adapter(adaptee)
cliente = Cliente(adapter)


# Fazendo uma requisição - Requisição Especifica do Adaptee:
cliente.criaRequest()

